📖 Add note about clusterctl upgrade test coverage

sbueringer opened the pull request:
Signed-off-by: Stefan Büringer buringerst@vmware.com

<!-- please add an icon to the title of this PR (see https://sigs.k8s.io/cluster-api/CONTRIBUTING.md#contributing-a-patch), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #8038

sbueringer commented:
/assign @fabriziopandini @CecileRobertMichon

CecileRobertMichon commented on docs/book/src/clusterctl/commands/upgrade.md:
```diff
@@ -75,6 +75,16 @@ clusterctl upgrade apply \
+
+Cluster API only tests clusterctl upgrades from the latest minor version of each contract for a specific version 
+(e.g. v0.3 (v1alpha3) => v1.3, v0.4 (v1alpha4) => v1.3 and v1.2 (v1beta1) => v1.3). While we are not testing skip-level 
+upgrades from earlier versions (e.g. v1.1 => v1.3), it should work in general.
```
> it should work in general

I wonder if this sounds a bit too vague, but I don't have better suggestions. Maybe others can weigh in?

sbueringer replied:
@CecileRobertMichon Kilian suggested an improvement. Does that sound better to you?

fabriziopandini replied:
+1 to the second version
Wondering if we can simplify the first sentence by just listing what upgrade path we are currently testing; also, if I'm not wrong we are missing 1.2 --> 1.3

sbueringer replied:
We can add an explicit list, but that also means we have to keep the list up-to-date.

I would prefer writing it more generic than that to avoid this

killianmuldoon commented on docs/book/src/clusterctl/commands/upgrade.md:
```diff
@@ -75,6 +75,16 @@ clusterctl upgrade apply \
 
+<h1>Skip-level upgrades</h1>
+
+Cluster API only tests clusterctl upgrades from the latest minor version of each contract for a specific version 
+(e.g. v0.3 (v1alpha3) => v1.3, v0.4 (v1alpha4) => v1.3 and v1.2 (v1beta1) => v1.3). While we are not testing skip-level 
+upgrades from earlier versions (e.g. v1.1 => v1.3), it should work in general.
```
Maybe?
```suggestion
Cluster API only tests clusterctl upgrades from the latest minor version of each contract for a specific version 
(e.g. v0.3 (v1alpha3) => v1.3, v0.4 (v1alpha4) => v1.3 and v1.2 (v1beta1) => v1.3). Skip-level upgrades from earlier versions (e.g. v1.1 => v1.3) are not blocked, but are not tested. Users intending to carry out skip-level upgrades should do their own validation to ensure the operation works correctly.
```

sbueringer replied:
Thx. I like it!

killianmuldoon commented:
/lgtm
with the updated text

chrischdi commented on docs/book/src/clusterctl/commands/upgrade.md:
```diff
@@ -75,6 +75,21 @@ clusterctl upgrade apply \
 
 <aside class="note warning">
 
+<h1>Skip-level upgrades</h1>
```
Should we call this "Skipping minor versions in upgrades"?

oscr replied:
Could we also have an upgrade or link in the [tasks / upgrading clusters](https://github.com/kubernetes-sigs/cluster-api/blob/329753cae2c23d0f29a764870dc99972a50036b0/docs/book/src/tasks/upgrading-clusters.md?plain=1#L13-L14) to this?

sbueringer replied:
@oscr I think this is a different topic. This PR is about skipping Cluster API minor versions while what you referenced is about skipping Kubernetes minor versions.

sbueringer commented:
@CecileRobertMichon @killianmuldoon @chrischdi @fabriziopandini 

We realized that we are actually testing a skip-level upgrade (the one from v1.2 => latest, as v1.2 is a supported release).
I rewrote the note and added a table to make the underlying idea more understandable

killianmuldoon commented on docs/book/src/clusterctl/commands/upgrade.md:
```diff
@@ -75,6 +77,26 @@ clusterctl upgrade apply \
+<h1>Clusterctl upgrade test coverage</h1>
+
+Cluster API only tests a subset of possible clusterctl upgrade paths as otherwise the test matrix would be overwhelming.
+Untested upgrade paths are not blocked by clusterctl and should work in general, they are just not tested. Users 
```
```suggestion
Untested upgrade paths are not blocked by clusterctl but they are just not tested. Users 
```

Nit, but I'm not sure about the grey area of "it should work, but it's not tested". Better IMO just to describe the state.

sbueringer replied:
Let's see what others thing. I think the assurance that this is something that we expect to work out-of-the-box might be important to users.

fabriziopandini replied:
We can keep it as it is, given that the most important part is the next part about users doing their own validation if they want to rely on untested upgrade paths

fabriziopandini reviewed (commented):
/lgtm
/approve
